### ISSUE TITLE
Union overload types not properly matched

### ISSUE THREAD

**Issue by Rapptz** [microsoft/pyright#1772]
**Describe the bug**
This one's a bit strange to explain. When overloading based off of the different union member types, passing a union as a parameter to that function will report an error that no such overload is found.

**To Reproduce**

```py
@overload
def parse_time(timestamp: None) -> None:
    ...

@overload
def parse_time(timestamp: str) -> datetime.datetime:
    ...

def parse_time(timestamp: Optional[str]) -> Optional[datetime.datetime]:
    if timestamp:
        return datetime.datetime.fromisoformat(timestamp)
    return None

def args(a: Optional[str]):
    b = parse_time(a)  # error
```

**Expected behavior**
Overload to be detected

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/1695103/115148390-5978e600-a02d-11eb-8a0a-e798760d5eb9.png)

**VS Code extension or command-line**
Pylance 2021.4.1

**Additional context**
I originally thought this was an issue with just `Optional` but reproducing the issue with a `Union` is possible as well:

```py
@overload
def test(x: str) -> int:
    ...

@overload
def test(x: int) -> str:
    ...

def test(x: Union[str, int]) -> Union[str, int]:
    if isinstance(x, str):
        return 10
    return 'hello'

def inner(b: Union[str, int]):
    x = test(b) # error
```

If I litter `reveal_type` in both the functions, mypy seems to correctly infer the types:
```
test.py:21: note: Revealed type is 'Union[datetime.datetime, None]'
test.py:38: note: Revealed type is 'Union[builtins.int, builtins.str]'
```

**Comment by erictraut** [comment]
This behavior is intended. See https://github.com/microsoft/pyright/issues/1565.

**Comment by Rapptz** [comment]
That's unfortunate. Not really sure how to work around this since it throws a wrench into this common pattern of mine for monadic functions. I guess I can make a 3rd overload but if it's just a duplicate of the actual declaration it looks really silly.

**Comment by erictraut** [comment]
This issue has been reported often enough that I've decided to change pyright's overload matching behavior to more closely match that of mypy's. If necessary, it now expands argument union types, potentially matching multiple overloads. If the union expansion results more than 64 combinations, it gives up and fails the call. In practice, this will be very rare, but it's something that the type evaluation engine needs to guard against.

This change will be included in the next release of pyright and pylance.

**Comment by erictraut** [comment]
This change is included in pyright 1.1.135, which I just published. It will also be included in the next release of pylance.
